# 1. Import Libraries and connect to PostgreSQL 

In [1]:
!pip install psycopg2-binary sqlalchemy

In [2]:
!pip install ipython-sql

In [3]:
%load_ext sql

In [4]:
%sql postgresql://postgres:171004@localhost:5432/telco_churn

In [5]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

In [7]:
%sql select * from customer_churn limit 10 ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
10 rows affected.


[('19877-DEL', 'Male', 35, 'No', 'Delhi', 7, 27, None, 'Yes', 'No', 'Yes', 'Cable', 'No', 'Yes', 'No', 'Yes', 'Yes', 'No', 'No', 'Yes', 'One Year', 'Yes', 'Credit Card', 65.6, 593.3, 0.0, 0.0, 381.51, 974.81, 'Stayed', None, None),
 ('58353-MAH', 'Female', 45, 'Yes', 'Maharashtra', 14, 13, None, 'Yes', 'Yes', 'Yes', 'Cable', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Yes', 'No', 'Month-to-Month', 'No', 'Credit Card', -4.0, 542.4, 38.33, 10.0, 96.21, 610.28, 'Stayed', None, None),
 ('25063-WES', 'Male', 51, 'No', 'West Bengal', 4, 35, 'Deal 5', 'Yes', 'No', 'Yes', 'Fiber Optic', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'Yes', 'Month-to-Month', 'Yes', 'Bank Withdrawal', 73.9, 280.85, 0.0, 0.0, 134.6, 415.45, 'Churned', 'Competitor', 'Competitor had better devices'),
 ('59787-KAR', 'Male', 79, 'No', 'Karnataka', 3, 21, 'Deal 4', 'Yes', 'No', 'Yes', 'Fiber Optic', 'No', 'Yes', 'Yes', 'No', 'Yes', 'Yes', 'No', 'Yes', 'Month-to-Month', 'Yes', 'Bank Withdrawal', 98.0, 1237.85, 0.0, 0.0, 361.66, 1599.51, 'Churned', 'Dissatisfaction', 'Product dissatisfaction'),
 ('28544-TAM', 'Female', 80, 'No', 'Tamil Nadu', 3, 8, None, 'Yes', 'No', 'Yes', 'Fiber Optic', 'No', 'No', 'No', 'Yes', 'Yes', 'No', 'No', 'Yes', 'Month-to-Month', 'Yes', 'Credit Card', 83.9, 267.4, 0.0, 0.0, 22.14, 289.54, 'Churned', 'Dissatisfaction', 'Network reliability'),
 ('32750-TEL', 'Male', 24, 'No', 'Telangana', 9, 2, 'Deal 5', 'Yes', 'No', 'Yes', 'Cable', 'No', 'No', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Month-to-Month', 'Yes', 'Credit Card', 69.4, 571.45, 0.0, 0.0, 150.93, 722.38, 'Stayed', None, None),
 ('84741-MAH', 'Female', 66, 'Yes', 'Maharashtra', 9, 23, 'Deal 1', 'Yes', 'No', 'Yes', 'Fiber Optic', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Two Year', 'Yes', 'Bank Withdrawal', 109.7, 7904.25, 0.0, 0.0, 707.16, 8611.41, 'Stayed', None, None),
 ('65618-GUJ', 'Male', 51, 'Yes', 'Gujarat', 12, 24, 'Deal 2', 'Yes', 'Yes', 'Yes', 'Fiber Optic', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'Two Year', 'Yes', 'Credit Card', 84.65, 5377.8, 0.0, 20.0, 816.48, 6214.28, 'Stayed', None, None),
 ('64233-GUJ', 'Female', 63, 'No', 'Gujarat', 4, 10, 'Deal 5', 'Yes', 'No', 'Yes', 'DSL', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Two Year', 'Yes', 'Bank Withdrawal', 48.2, 340.35, 0.0, 0.0, 73.71, 414.06, 'Stayed', None, None),
 ('23612-UTT', 'Female', 41, 'No', 'Uttar Pradesh', 3, 1, None, 'Yes', 'Yes', 'Yes', 'Cable', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Two Year', 'Yes', 'Credit Card', 90.45, 5957.9, 0.0, 0.0, 1849.9, 7807.8, 'Stayed', None, None)]

# 2. Data cleaning 

### Count null values of each columns 

In [8]:
%%sql
select 
 count (*) filter (where customer_id is null) as customerid_null , 
 count (*) filter (where gender is null) as gender_null , 
 count (*) filter (where age is null) as age_null ,
 count (*) filter (where married is null) as married_null ,
 count (*) filter (where state is null) as state_null ,
 count (*) filter (where number_of_referrals is null) as referrals_null ,
 count (*) filter (where tenure_in_months is null) as tenure_null ,
 count (*) filter (where value_deal is null) as value_deal_null ,
 count (*) filter (where phone_service is null) as phone_service_null ,
 count (*) filter (where multiple_lines is null) as multiple_lines_null ,
 count (*) filter (where internet_service is null) as internet_service_null ,
 count (*) filter (where internet_type is null) as internet_type_null ,
 count (*) filter (where online_security is null) as online_security_null ,
 count (*) filter (where online_backup is null) as online_backup_null ,
 count (*) filter (where device_protection_plan is null) as deviceProtect_null ,
 count (*) filter (where premium_support is null) as preSp_null ,
 count (*) filter (where streaming_tv is null) as streamingTV_null ,
 count (*) filter (where streaming_movies is null) as stream_movies_null ,
 count (*) filter (where streaming_music is null) as stream_music_null ,
 count (*) filter (where unlimited_data is null) as unlimited_null ,
 count (*) filter (where contract is null) as contract_null ,
 count (*) filter (where paperless_billing is null) as paperlessbill_null ,
 count (*) filter (where payment_method is null) as payMethod_null ,
 count (*) filter (where monthly_charge is null) as monthly_charge_null ,
 count (*) filter (where total_charges is null) as total_charges_null ,
 count (*) filter (where total_refunds is null) as total_refunds_null ,
 count (*) filter (where total_extra_data_charges is null) as total_extra_charges_null ,
 count (*) filter (where total_long_distance_charges is null) as total_long_distance_charges_null ,
 count (*) filter (where total_revenue is null) as total_revenue_null ,
 count (*) filter (where customer_status is null) as customer_status_null ,
 count (*) filter (where churn_category is null) as churn_category_null ,
 count (*) filter (where churn_reason is null) as churn_reason_null 

from customer_churn ; 


 * postgresql://postgres:***@localhost:5432/telco_churn
1 rows affected.


[(0, 0, 0, 0, 0, 0, 0, 3548, 0, 622, 0, 1390, 1390, 1390, 1390, 1390, 1390, 1390, 1390, 1390, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4686, 4686)]

### After observing the result , we continue to process data in the following direction:  
value_deal : 3548 -> 'No deal' <br>
multiple_lines : 622  -> 'No' <br>
internet_type : 1390  -> 'None' <br> 
online_security : 1390 -> 'No' <br>
online_backup: 1390 -> 'No'<br>
device_protection_plan : 1390  -> 'No' <br>
premium_support : 1390 -> 'No'<br>
streaming_tv : 1390 -> 'No' <br>
streaming_movies : 1390 -> 'No' <br>
streaming_music : 1390 -> 'No'<br>
unlimited_data : 1390  -> 'No' <br>
churn_category : 4686  -> 'Others'<br>
churn_reason : 4686  -> 'Others'<br> 

In [15]:
%config SqlMagic.autocommit=True


In [16]:
# Update null values 
%%sql 
update customer_churn 
set 
  value_deal = coalesce(value_deal, 'No deal') , 
  multiple_lines = COALESCE(multiple_lines, 'No'),
  internet_type = COALESCE(internet_type, 'None'),
  online_security = COALESCE(online_security, 'No'),
  online_backup = COALESCE(online_backup, 'No'),
  device_protection_plan = COALESCE(device_protection_plan, 'No'),
  premium_support = COALESCE(premium_support, 'No'),
  streaming_tv = COALESCE(streaming_tv, 'No'),
  streaming_movies = COALESCE(streaming_movies, 'No'),
  streaming_music = COALESCE(streaming_music, 'No'),
  unlimited_data = COALESCE(unlimited_data, 'No'),
  churn_category = COALESCE(churn_category, 'Others'),
  churn_reason = COALESCE(churn_reason, 'Others')

 * postgresql://postgres:***@localhost:5432/telco_churn
6418 rows affected.


[]

In [17]:
# re-check 
%sql select * from customer_churn limit 5 ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
5 rows affected.


[('47587-KER', 'Female', 45, 'Yes', 'Kerala', 13, 5, 'No deal', 'Yes', 'No', 'No', 'None', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'One Year', 'No', 'Bank Withdrawal', -4.0, 395.6, 0.0, 0.0, 146.2, 541.8, 'Stayed', 'Others', 'Others'),
 ('53637-KER', 'Male', 22, 'No', 'Kerala', 10, 11, 'Deal 4', 'Yes', 'No', 'No', 'None', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'One Year', 'No', 'Credit Card', 20.65, 417.5, 0.0, 0.0, 929.4, 1346.9, 'Stayed', 'Others', 'Others'),
 ('84442-HAR', 'Female', 55, 'No', 'Haryana', 13, 20, 'Deal 2', 'Yes', 'No', 'Yes', 'Cable', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Yes', 'Yes', 'One Year', 'No', 'Credit Card', 55.65, 2688.85, 0.0, 0.0, 93.38, 2782.23, 'Stayed', 'Others', 'Others'),
 ('19877-DEL', 'Male', 35, 'No', 'Delhi', 7, 27, 'No deal', 'Yes', 'No', 'Yes', 'Cable', 'No', 'Yes', 'No', 'Yes', 'Yes', 'No', 'No', 'Yes', 'One Year', 'Yes', 'Credit Card', 65.6, 593.3, 0.0, 0.0, 381.51, 974.81, 'Stayed', 'Others', 'Others'),
 ('58353-MAH', 'Female', 45, 'Yes', 'Maharashtra', 14, 13, 'No deal', 'Yes', 'Yes', 'Yes', 'Cable', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Yes', 'No', 'Month-to-Month', 'No', 'Credit Card', -4.0, 542.4, 38.33, 10.0, 96.21, 610.28, 'Stayed', 'Others', 'Others')]

### We found some negative values in column 'monthly_charge' . Let's check how many such values there are in the whole dataset 

In [18]:
%%sql 
select 
 count(*) filter (where monthly_charge < 0) as monthly_charge, 
 count(*) filter (where total_charges < 0) as total_charges, 
 count(*) filter (where total_extra_data_charges < 0) as total_extra_data_charges, 
 count(*) filter (where total_long_distance_charges < 0) as total_long_distance_charges , 
 count(*) filter (where total_refunds < 0) as total_refund , 
 count(*) filter (where total_revenue < 0) as total_revenue
from customer_churn ;

 * postgresql://postgres:***@localhost:5432/telco_churn
1 rows affected.


[(107, 0, 0, 0, 0, 0)]

### There are 107 negative values in column 'monthly_charge' only , as this is a small amount , we decide to delete them all  

In [19]:
%%sql 
delete from customer_churn 
where monthly_charge < 0 ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
107 rows affected.


[]

### Add a new column 'Churn' from 'customer_status' 

In [20]:
%%sql 
alter table customer_churn 
add column churn float ; 

update customer_churn 
set 
 churn = case 
   when customer_status = 'Churned' then 1 
   else 0 
  end; 

 * postgresql://postgres:***@localhost:5432/telco_churn
Done.
6311 rows affected.


[]

### Add new column 'tenure_category' to distinguish between new customers , long-term customers , ... 

In [21]:
%%sql 
alter table customer_churn 
add column tenure_category text ; 

update customer_churn 
 set tenure_category = case 
   when tenure_in_months <= 12 then 'New customer'
   when tenure_in_months between 13 and 24 then 'Mid-term customer'
   when tenure_in_months > 24 then 'Long-term customer'
  end; 

 * postgresql://postgres:***@localhost:5432/telco_churn
Done.
6311 rows affected.


[]

# 3. QUERIES 

### 1. Sum of churns 


In [22]:
%%sql 
select sum(churn)
from customer_churn ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
1 rows affected.


[(1705.0,)]

### 2. Churn rate by gender 

In [23]:
%%sql 
select gender , count(customer_id) as total , round( sum(churn)::numeric/count(customer_id) , 3) as churn_rate , sum(churn)
from customer_churn 
group by gender;

 * postgresql://postgres:***@localhost:5432/telco_churn
2 rows affected.


[('Female', 3981, Decimal('0.275'), 1093.0),
 ('Male', 2330, Decimal('0.263'), 612.0)]

### 3. Churn Rate by Marital Status

In [24]:
%%sql 
select married , count(customer_id) as total , round( sum(churn)::numeric/count(customer_id) , 2) as churn_rate
from customer_churn
group by married ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
2 rows affected.


[('No', 3167, Decimal('0.27')), ('Yes', 3144, Decimal('0.27'))]

### 4. Churn rate by state 

In [25]:
%%sql 
select state , count(customer_id) as total , round( sum(churn)::numeric/count(customer_id) , 2) as churn_rate
from customer_churn
group by state
order by churn_rate desc ; 

 * postgresql://postgres:***@localhost:5432/telco_churn
22 rows affected.


[('Jammu & Kashmir', 314, Decimal('0.57')),
 ('Assam', 134, Decimal('0.39')),
 ('Jharkhand', 112, Decimal('0.34')),
 ('Chhattisgarh', 56, Decimal('0.32')),
 ('Odisha', 149, Decimal('0.30')),
 ('Delhi', 125, Decimal('0.30')),
 ('Puducherry', 41, Decimal('0.27')),
 ('Kerala', 197, Decimal('0.27')),
 ('Rajasthan', 255, Decimal('0.27')),
 ('West Bengal', 365, Decimal('0.26')),
 ('Bihar', 330, Decimal('0.26')),
 ('Tamil Nadu', 587, Decimal('0.26')),
 ('Haryana', 390, Decimal('0.26')),
 ('Andhra Pradesh', 390, Decimal('0.25')),
 ('Telangana', 277, Decimal('0.25')),
 ('Uttar Pradesh', 619, Decimal('0.25')),
 ('Punjab', 338, Decimal('0.23')),
 ('Maharashtra', 494, Decimal('0.23')),
 ('Uttarakhand', 61, Decimal('0.23')),
 ('Karnataka', 463, Decimal('0.23')),
 ('Gujarat', 329, Decimal('0.22')),
 ('Madhya Pradesh', 285, Decimal('0.22'))]

### 5. Top 5 states which got the most referrals to our company 

In [26]:
%%sql 
select state , count(number_of_referrals) as total_referrals 
from customer_churn 
group by state 
order by total_referrals desc 
limit 5 ;

 * postgresql://postgres:***@localhost:5432/telco_churn
5 rows affected.


[('Uttar Pradesh', 619),
 ('Tamil Nadu', 587),
 ('Maharashtra', 494),
 ('Karnataka', 463),
 ('Haryana', 390)]

### 6. Churn rate by customer loyalty 

In [27]:
%%sql 
select tenure_category , count(customer_id) as total , round( sum(churn)::numeric/count(customer_id) , 2) as churn_rate
from customer_churn
group by tenure_category
order by churn_rate desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
3 rows affected.


[('Long-term customer', 1907, Decimal('0.28')),
 ('New customer', 2515, Decimal('0.27')),
 ('Mid-term customer', 1889, Decimal('0.26'))]

### 7. Churn rate by value deal 

In [28]:
%%sql 
select value_deal , count(customer_id) as total , round( sum(churn)::numeric/count(customer_id) , 2) as churn_rate
from customer_churn
group by value_deal
order by churn_rate desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
6 rows affected.


[('Deal 5', 725, Decimal('0.54')),
 ('Deal 4', 531, Decimal('0.27')),
 ('No deal', 3481, Decimal('0.27')),
 ('Deal 3', 361, Decimal('0.22')),
 ('Deal 2', 750, Decimal('0.13')),
 ('Deal 1', 463, Decimal('0.08'))]

### 8. Churn rate by service groups : 

In [29]:
%%sql 
select 'phone_service' as service, phone_service as status,
       count(customer_id) as total,
       round(sum(churn)::numeric / count(customer_id), 2) as churn_rate
from customer_churn
group by phone_service

union all
select 'multiple_lines', multiple_lines,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by multiple_lines

union all
select 'internet_service', internet_service,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by internet_service

union all
select 'online_security', online_security,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by online_security

union all
select 'online_backup', online_backup,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by online_backup

union all
select 'device_protection_plan', device_protection_plan,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by device_protection_plan

union all
select 'premium_support', premium_support,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by premium_support

union all
select 'streaming_tv', streaming_tv,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by streaming_tv

union all
select 'streaming_movies', streaming_movies,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by streaming_movies

union all
select 'streaming_music', streaming_music,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by streaming_music

union all
select 'unlimited_data', unlimited_data,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by unlimited_data

union all
select 'paperless_billing', paperless_billing,
       count(customer_id),
       round(sum(churn)::numeric / count(customer_id), 2)
from customer_churn
group by paperless_billing;


 * postgresql://postgres:***@localhost:5432/telco_churn
24 rows affected.


[('phone_service', 'No', 614, Decimal('0.26')),
 ('phone_service', 'Yes', 5697, Decimal('0.27')),
 ('paperless_billing', 'No', 2578, Decimal('0.17')),
 ('paperless_billing', 'Yes', 3733, Decimal('0.34')),
 ('unlimited_data', 'No', 2048, Decimal('0.17')),
 ('unlimited_data', 'Yes', 4263, Decimal('0.32')),
 ('streaming_music', 'No', 4086, Decimal('0.25')),
 ('streaming_music', 'Yes', 2225, Decimal('0.30')),
 ('streaming_movies', 'No', 3842, Decimal('0.25')),
 ('streaming_movies', 'Yes', 2469, Decimal('0.31')),
 ('streaming_tv', 'No', 3875, Decimal('0.25')),
 ('streaming_tv', 'Yes', 2436, Decimal('0.31')),
 ('premium_support', 'No', 4489, Decimal('0.32')),
 ('premium_support', 'Yes', 1822, Decimal('0.16')),
 ('device_protection_plan', 'No', 4134, Decimal('0.29')),
 ('device_protection_plan', 'Yes', 2177, Decimal('0.23')),
 ('online_backup', 'No', 4138, Decimal('0.29')),
 ('online_backup', 'Yes', 2173, Decimal('0.22')),
 ('online_security', 'No', 4518, Decimal('0.32')),
 ('online_security', 'Yes', 1793, Decimal('0.15')),
 ('internet_service', 'No', 1366, Decimal('0.08')),
 ('internet_service', 'Yes', 4945, Decimal('0.32')),
 ('multiple_lines', 'No', 3640, Decimal('0.26')),
 ('multiple_lines', 'Yes', 2671, Decimal('0.29'))]

### 9. Churn rate by internet type

In [30]:
%%sql 
select internet_type , count(customer_id) , round(sum(churn)::numeric / count(customer_id), 2) as churn_rate
from customer_churn
group by internet_type
order by churn_rate desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
4 rows affected.


[('Fiber Optic', 2725, Decimal('0.41')),
 ('Cable', 750, Decimal('0.26')),
 ('DSL', 1470, Decimal('0.19')),
 ('None', 1366, Decimal('0.08'))]

### 10. Churn rate by contract 

In [31]:
%%sql 
select contract , count(customer_id) , round(sum(churn)::numeric / count(customer_id), 2) as churn_rate
from customer_churn
group by contract
order by churn_rate desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
3 rows affected.


[('Month-to-Month', 3234, Decimal('0.46')),
 ('One Year', 1386, Decimal('0.11')),
 ('Two Year', 1691, Decimal('0.03'))]

### 11. Churn rate by payment method

In [32]:
%%sql 
select payment_method , count(customer_id) , round(sum(churn)::numeric / count(customer_id), 2) as churn_rate
from customer_churn
group by payment_method
order by churn_rate desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
3 rows affected.


[('Mailed Check', 343, Decimal('0.38')),
 ('Bank Withdrawal', 3521, Decimal('0.35')),
 ('Credit Card', 2447, Decimal('0.15'))]

### 12. Total churns of each churn category  

In [33]:
%%sql 
select churn_category , sum(churn) as total
from customer_churn
group by churn_category
order by total desc ;

 * postgresql://postgres:***@localhost:5432/telco_churn
6 rows affected.


[('Competitor', 752.0),
 ('Attitude', 294.0),
 ('Dissatisfaction', 293.0),
 ('Price', 194.0),
 ('Other', 172.0),
 ('Others', 0.0)]

# 4. Summary of Findings

- Churn rate does not differ between gender and marital status.  

- The state with the highest Churn rate is **Jammu & Kashmir** (0.57), while **Uttar Pradesh** has the most Referrals (4682).  

- For the group of reasons for leaving the service:  
  - The majority of customers did not leave a specific reason (4778).  
  - The most common explicit reason is **Competitor** (752).  

- In the following service groups, the Churn rate of the **registered group** is higher than the non-registered group:
  - Internet Service  
  - Multiple Lines (multiple phone numbers at the same time)  
  - Paperless Billing (electronic bill)  
  - Phone Service  
  - Streaming Movies  
  - Streaming Music  
  - Streaming TV  
  - Unlimited Data  

- On the contrary, for the following service groups, the Churn rate of the **non-registered group** is higher than the registered group:
  - Device Protection Plan  
  - Online Backup  
  - Online Security  
  - Premium Support  

- For Age groups:  
  - The majority are from **20–59 years old**.  
  - The highest Churn rate is **0.36** in the elderly group (≥ 60).  

- For promotional packages:  
  - The Churn rate increases gradually with the number of Deals.  
  - The higher the Deal, the higher the Churn rate.  
  - The highest is **Deal 5** with a Churn rate of **0.54**.  

- According to the Tenure classification:  
  - The **New customers** group (≤ 12 months) has the highest number.  
  - The Churn rate across all 3 Tenure groups is quite similar.  
  - Surprisingly, the **Long-term customers** group (> 24 months) has the highest Churn rate (**0.28**).  

- Regarding the Internet network classification:  
  - The most used type is **Fiber Optic** (2725).  
  - It also has the highest Churn rate (**0.41**).  

- Regarding the Contract type:  
  - **Month-to-month** contracts account for the majority and also have the highest Churn rate (**0.46**).  
  - The Churn rate decreases as the contract term increases, which is reasonable.  

- Regarding the Payment Method:  
  - The most popular is **Bank Withdrawal** (3521).  
  - The highest Churn rate is **Mailed Check** (0.38) with only 343 customers.  
  - **Credit Card** has many customers (2447) but the lowest Churn rate (**0.15**).  
